File: macos/Sources/ElevateCore/Coordination/ActivationCoordinator.swift
import Foundation

public struct ActivationOutcome: Hashable, Sendable {
    public enum Result: Hashable, Sendable {
        case activated(ActiveAssignment)
        case pendingApproval(ActiveAssignment)
        case scheduled(ActiveAssignment)
        case failed(PIMError)
    }
    public let roleKey: RoleKey
    public let result: Result
    public init(roleKey: RoleKey, result: Result) { self.roleKey = roleKey; self.result = result }
}

/// Runs activations grouped by identity+tenant: groups in parallel, requests within a group in sequence.
/// Handles `interactionRequired` and `claimsChallenge` with one interactive prompt and one retry per request.
public final class ActivationCoordinator: Sendable {
    private let providers: [RoleScopeKind: any PIMProvider]
    private let tokens: any TokenProviding

    public init(providers: [any PIMProvider], tokens: any TokenProviding) {
        self.providers = Dictionary(uniqueKeysWithValues: providers.map { ($0.kind, $0) })
        self.tokens = tokens
    }

    public func provider(for kind: RoleScopeKind) -> (any PIMProvider)? { providers[kind] }

    public func activate(_ requests: [ActivationRequest], identities: [Identity],
                         onProgress: @Sendable @escaping (ActivationOutcome) -> Void) async -> [ActivationOutcome] {
        let identityById = Dictionary(uniqueKeysWithValues: identities.map { ($0.id, $0) })
        let groups = Dictionary(grouping: requests) { $0.roleKey.tenantKey }
        return await withTaskGroup(of: [ActivationOutcome].self) { group in
            for (tenantKey, groupRequests) in groups {
                group.addTask { [self] in
                    var outcomes: [ActivationOutcome] = []
                    for request in groupRequests {
                        let outcome = await self.activateOne(request, identity: identityById[tenantKey.identityId])
                        onProgress(outcome)
                        outcomes.append(outcome)
                    }
                    return outcomes
                }
            }
            var all: [ActivationOutcome] = []
            for await batch in group { all += batch }
            return all
        }
    }

    private func activateOne(_ request: ActivationRequest, identity: Identity?) async -> ActivationOutcome {
        guard let identity else {
            return ActivationOutcome(roleKey: request.roleKey, result: .failed(.unexpected(status: 0, body: "Unknown identity \(request.roleKey.identityId)")))
        }
        guard let provider = providers[request.roleKey.scope.kind] else {
            return ActivationOutcome(roleKey: request.roleKey, result: .failed(.unexpected(status: 501, body: "No provider for \(request.roleKey.scope.kind)")))
        }
        // A role can demand a fresh MFA or a Conditional Access authentication context. The
        // service answers with a 400 and no claims header, so we choose the claims: the policy's
        // authentication context when it has one, otherwise a plain MFA re-verification.
        let fallbackClaims = request.authenticationContext.map(ClaimsChallenge.authenticationContext) ?? ClaimsChallenge.multiFactor
        do {
            let assignment = try await InteractionRetry.run(tokens: tokens, identity: identity, tenantId: request.roleKey.tenantId,
                                                            scopes: provider.scopes, fallbackClaims: fallbackClaims) {
                try await provider.activate(request, identity: identity)
            }
            switch assignment.status {
            case .pendingApproval: return ActivationOutcome(roleKey: request.roleKey, result: .pendingApproval(assignment))
            case .scheduled: return ActivationOutcome(roleKey: request.roleKey, result: .scheduled(assignment))
            case .failed(let m): return ActivationOutcome(roleKey: request.roleKey, result: .failed(.unexpected(status: 0, body: m)))
            default: return ActivationOutcome(roleKey: request.roleKey, result: .activated(assignment))
            }
        } catch PIMError.interactionRequired, PIMError.claimsChallenge {
            // The retry already sent the user through the browser once; a second refusal means
            // that sign-in did not satisfy the role's requirement.
            let requirement = request.authenticationContext.map { "the Conditional Access authentication context \"\($0)\"" } ?? "multi-factor authentication"
            return ActivationOutcome(roleKey: request.roleKey, result: .failed(.policyViolation(
                "This role requires \(requirement) and the sign-in did not satisfy it. Try again and complete the verification in the browser.")))
        } catch let e as PIMError {
            return ActivationOutcome(roleKey: request.roleKey, result: .failed(e))
        } catch {
            return ActivationOutcome(roleKey: request.roleKey, result: .failed(.network(error.localizedDescription)))
        }
    }

    public func deactivate(_ assignment: ActiveAssignment, identity: Identity) async throws {
        guard let provider = providers[assignment.roleKey.scope.kind] else {
            throw PIMError.unexpected(status: 501, body: "No provider for \(assignment.roleKey.scope.kind)")
        }
        try await InteractionRetry.run(tokens: tokens, identity: identity, tenantId: assignment.roleKey.tenantId, scopes: provider.scopes) {
            try await provider.deactivate(assignment, identity: identity)
        }
    }

    public func cancelPendingRequest(_ assignment: ActiveAssignment, identity: Identity) async throws {
        guard let provider = providers[assignment.roleKey.scope.kind] else {
            throw PIMError.unexpected(status: 501, body: "No provider for \(assignment.roleKey.scope.kind)")
        }
        try await InteractionRetry.run(tokens: tokens, identity: identity, tenantId: assignment.roleKey.tenantId, scopes: provider.scopes) {
            try await provider.cancelPendingRequest(assignment, identity: identity)
        }
    }
}

File: windows/src/Elevate.Core/Coordination/ActivationCoordinator.cs
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;
using Elevate.Core.Providers;

namespace Elevate.Core.Coordination;

/// <summary>What became of one activation. Port of Swift's <c>ActivationOutcome.Result</c>.</summary>
public abstract record ActivationResult
{
    private ActivationResult()
    {
    }

    public sealed record Activated(ActiveAssignment Assignment) : ActivationResult;

    public sealed record PendingApproval(ActiveAssignment Assignment) : ActivationResult;

    public sealed record Scheduled(ActiveAssignment Assignment) : ActivationResult;

    public sealed record Failed(PimException Error) : ActivationResult;
}

public sealed record ActivationOutcome(RoleKey RoleKey, ActivationResult Result);

/// <summary>
/// Runs activations grouped by identity+tenant: groups in parallel, requests within a group in sequence.
/// Handles <see cref="PimErrorKind.InteractionRequired"/> and <see cref="PimErrorKind.ClaimsChallenge"/>
/// with one interactive prompt and one retry per request.
/// </summary>
public sealed class ActivationCoordinator
{
    private readonly Dictionary<RoleScopeKind, IPimProvider> _providers;
    private readonly ITokenProvider _tokens;

    public ActivationCoordinator(IEnumerable<IPimProvider> providers, ITokenProvider tokens)
    {
        ArgumentNullException.ThrowIfNull(providers);
        _providers = providers.ToDictionary(p => p.Kind);
        _tokens = tokens;
    }

    public IPimProvider? Provider(RoleScopeKind kind) => _providers.GetValueOrDefault(kind);

    public async Task<IReadOnlyList<ActivationOutcome>> ActivateAsync(
        IEnumerable<ActivationRequest> requests,
        IEnumerable<Identity> identities,
        Action<ActivationOutcome>? onProgress = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(requests);
        ArgumentNullException.ThrowIfNull(identities);

        var identityById = identities.ToDictionary(i => i.Id);
        var groups = requests.GroupBy(r => r.RoleKey.TenantKey).ToList();

        var batches = await Task.WhenAll(groups.Select(group => Task.Run(async () =>
        {
            var outcomes = new List<ActivationOutcome>();
            foreach (var request in group)
            {
                var outcome = await ActivateOneAsync(request, identityById.GetValueOrDefault(group.Key.IdentityId), ct)
                    .ConfigureAwait(false);
                onProgress?.Invoke(outcome);
                outcomes.Add(outcome);
            }

            return outcomes;
        }, ct))).ConfigureAwait(false);

        return [.. batches.SelectMany(b => b)];
    }

    private async Task<ActivationOutcome> ActivateOneAsync(ActivationRequest request, Identity? identity, CancellationToken ct)
    {
        if (identity is null)
        {
            return Failure(request.RoleKey, new PimException(
                PimErrorKind.Unexpected, $"Unknown identity {request.RoleKey.IdentityId}"));
        }

        if (Provider(request.RoleKey.Scope.Kind) is not { } provider)
        {
            return Failure(request.RoleKey, new PimException(
                PimErrorKind.Unexpected, $"No provider for {CaseName(request.RoleKey.Scope.Kind)}", 501));
        }

        // A role can demand a fresh MFA or a Conditional Access authentication context. The
        // service answers with a 400 and no claims header, so we choose the claims: the policy's
        // authentication context when it has one, otherwise a plain MFA re-verification.
        var fallbackClaims = request.AuthenticationContext is { } context
            ? ClaimsChallenge.AuthenticationContext(context)
            : ClaimsChallenge.MultiFactor;

        try
        {
            var assignment = await InteractionRetry.RunAsync(
                _tokens, identity, request.RoleKey.TenantId, provider.Scopes,
                () => provider.ActivateAsync(request, identity, ct), fallbackClaims, ct).ConfigureAwait(false);

            return new ActivationOutcome(request.RoleKey, assignment.Status.Kind switch
            {
                AssignmentStatusKind.PendingApproval => new ActivationResult.PendingApproval(assignment),
                AssignmentStatusKind.Scheduled => new ActivationResult.Scheduled(assignment),
                AssignmentStatusKind.Failed => new ActivationResult.Failed(
                    new PimException(PimErrorKind.Unexpected, assignment.Status.FailureReason)),
                _ => new ActivationResult.Activated(assignment),
            });
        }
        catch (PimException e) when (e.Kind is PimErrorKind.InteractionRequired or PimErrorKind.ClaimsChallenge)
        {
            // The retry already sent the user through the browser once; a second refusal means
            // that sign-in did not satisfy the role's requirement.
            var requirement = request.AuthenticationContext is { } acrs
                ? $"the Conditional Access authentication context \"{acrs}\""
                : "multi-factor authentication";
            return Failure(request.RoleKey, new PimException(
                PimErrorKind.PolicyViolation,
                $"This role requires {requirement} and the sign-in did not satisfy it. "
                + "Try again and complete the verification in the browser."));
        }
        catch (PimException e)
        {
            return Failure(request.RoleKey, e);
        }
        catch (Exception e) when (e is not OperationCanceledException)
        {
            return Failure(request.RoleKey, new PimException(PimErrorKind.Network, e.Message));
        }
    }

    public async Task DeactivateAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);
        var provider = RequireProvider(assignment.RoleKey.Scope.Kind);
        await InteractionRetry.RunAsync(
            _tokens, identity, assignment.RoleKey.TenantId, provider.Scopes,
            async () =>
            {
                await provider.DeactivateAsync(assignment, identity, ct).ConfigureAwait(false);
                return true;
            },
            ct: ct).ConfigureAwait(false);
    }

    public async Task CancelPendingRequestAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);
        var provider = RequireProvider(assignment.RoleKey.Scope.Kind);
        await InteractionRetry.RunAsync(
            _tokens, identity, assignment.RoleKey.TenantId, provider.Scopes,
            async () =>
            {
                await provider.CancelPendingRequestAsync(assignment, identity, ct).ConfigureAwait(false);
                return true;
            },
            ct: ct).ConfigureAwait(false);
    }

    private IPimProvider RequireProvider(RoleScopeKind kind) =>
        Provider(kind) ?? throw new PimException(PimErrorKind.Unexpected, $"No provider for {CaseName(kind)}", 501);

    private static ActivationOutcome Failure(RoleKey key, PimException error) =>
        new(key, new ActivationResult.Failed(error));

    /// <summary>The Swift case name of a scope kind, so messages read the same on both platforms.</summary>
    private static string CaseName(RoleScopeKind kind)
    {
        var name = kind.ToString();
        return char.ToLowerInvariant(name[0]) + name[1..];
    }
}
